tfjs-layers/src/layers/nlp/modeling/transformer_decoder_test.ts

```diff
@@ -0,0 +1,177 @@
+import { input, model } from '../../../exports';
+import { expectTensorsClose } from '../../../utils/test_utils';
+import { Dense } from '../../core';
+import { sliceUpdate } from '../utils';
+
+import { TransformerDecoder } from './transformer_decoder';
+
+describe('TransformerDecoder', () => {
+  let originalTimeout: number;
+
+  beforeAll(() => {
+    // This test needs more time to finish the async fetch, adjusting
+    // jasmine timeout for this test to avoid flakiness. See jasmine
+    // documentation for detail:
+    // https://jasmine.github.io/2.0/introduction.html#section-42
+    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000000;
+  });
+
+  afterAll(() => jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout);
+
+  describe('valid call', () => {
+    function testValidCallWithoutCrossAttention(
+        testcaseName: string, normalizeFirst: boolean) {
+      it(`${testcaseName} without cross attention`, () => {
+        const decoderInput = randomUniform([4, 6]);
+        const decoder = new TransformerDecoder({
+          intermediateDim: 4,
+          numHeads: 2,
+          normalizeFirst,
+        });
+        expect(() => decoder.apply(decoderInput)).not.toThrow();
+      });
+    }
+
+    const params: Array<[string, boolean]> = [
+      ['without_norm_first', false],
+      ['with_norm_first', true],
+    ];
+
+    for (const [testcaseName, normalizeFirst] of params) {
+      testValidCallWithoutCrossAttention(testcaseName, normalizeFirst);
+    }
+  });
+
+  it('invalid call', () => {
+    const encoderInput = zeros([4, 6]);
+    const decoderInput = zeros([4, 6]);
+
+    // Without cross-attention.
+    const decoder = new TransformerDecoder({intermediateDim: 4, numHeads: 2});
+    decoder.apply(decoderInput);
+
+    // Should raise ValueError if encoderInput is provided.
+    expect(
+      () => decoder.apply(decoderInput, {encoderSequence: encoderInput})
+    ).toThrow();
+  });
+
+  it('error when invalid kernel initializer', () => {
+    expect(() => new TransformerDecoder({
+      intermediateDim: 4,
+      numHeads: 2,
+      dropout: 0.5,
+      kernelInitializer: 'Invalid',
+    })).toThrow();
+  });
+
+  it('one training step of transformer without cross attention', async () => {
+    const decoderInput = input({shape: [4, 6]});
+    const decoder = new TransformerDecoder({intermediateDim: 4, numHeads: 2});
+    let outputs = decoder.apply(decoderInput);
+    outputs = new Dense({
+      units: 10, activation: 'softmax'}).apply(outputs) as SymbolicTensor;
+    const tModel = model({inputs: decoderInput, outputs});
+
+    const decoderSequence = randomUniform([2, 4, 6]);
+    const label = randomUniformInt([2, 4, 1], 0, 10).asType('float32');
+
+    tModel.compile({loss: 'sparseCategoricalCrossentropy', optimizer: 'adam'});
+    const loss = tModel.trainOnBatch(decoderSequence, label);
+
+    expect(await loss).toBeGreaterThan(0);
+  });

```

Comment:
It's great that training works too (at least for this layer if not the whole model yet)!